src/simsoptpp/curveplanarfourier.cpp

```diff
@@ -0,0 +1,1557 @@
+#include "curveplanarfourier.h"
+
+template<class Array>
+double CurvePlanarFourier<Array>::inv_magnitude() {
+    double s = q[0] * q[0] + q[1] * q[1] + q[2] * q[2] + q[3] * q[3];
+    if(s != 0) {
+        return 1 / std::sqrt(s);
+    } else {
+        return 1;
+    }
+
+}
+
+template<class Array>
+void CurvePlanarFourier<Array>::gamma_impl(Array& data, Array& quadpoints) {
+    double sinphi, cosphi, siniphi, cosiphi;
+    int numquadpoints = quadpoints.size();
+    data *= 0;
+
+    Array q_norm = q * inv_magnitude();
+
+
+    for (int k = 0; k < numquadpoints; ++k) {
+        double phi = 2 * M_PI * quadpoints[k];
+        sinphi = sin(phi);
+        cosphi = cos(phi);
+        data(k, 0) = rc[0] * cosphi;
+        data(k, 1) = rc[0] * sinphi;
+        for (int i = 1; i < order+1; ++i) {
+            siniphi = sin(i * phi);
+            cosiphi = cos(i * phi);
+            data(k, 0) += (rc[i] * cosiphi + rs[i-1] * siniphi) * cosphi;
+            data(k, 1) += (rc[i] * cosiphi + rs[i-1] * siniphi) * sinphi;
+        }
```

Comment:
Look at how I refactored this loop to minimize the number of calls to `sin()` and `cos()`. These transcendental functions are slow, so it is good to call them as few times as possible, by storing the results in `cosphi`, `cosiphi`, etc. Please apply a similar refactoring to the other functions in this file to reduce the number of calls to `sin()` and `cos()`.